components/ai_chat/core/browser/text_embedder.cc

```diff
@@ -0,0 +1,309 @@
+/* Copyright (c) 2024 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/ai_chat/core/browser/text_embedder.h"
+
+#include <algorithm>
+
+#include "base/check_is_test.h"
+#include "base/files/file.h"
+#include "base/files/file_util.h"
+#include "base/hash/hash.h"
+#include "base/logging.h"
+#include "base/memory/ptr_util.h"
+#include "base/metrics/histogram_functions.h"
+#include "base/strings/strcat.h"
+#include "base/strings/string_split.h"
+#include "base/task/bind_post_task.h"
+#include "base/timer/elapsed_timer.h"
+#include "third_party/tflite/src/tensorflow/lite/core/api/op_resolver.h"
+#include "third_party/tflite_support/src/tensorflow_lite_support/cc/task/text/utils/text_op_resolver.h"
+
+using TFLiteTextEmbedder = tflite::task::text::TextEmbedder;
+using TFLiteTextEmbedderOptions = tflite::task::text::TextEmbedderOptions;
+using tflite::task::text::CreateTextOpResolver;
+
+namespace ai_chat {
+
+size_t TextEmbedder::g_segment_size_ceiling_(300);
+size_t TextEmbedder::g_segment_size_floor_(50);
+
+// static
+std::unique_ptr<TextEmbedder, base::OnTaskRunnerDeleter> TextEmbedder::Create(
+    const base::FilePath& model_path,
+    scoped_refptr<base::SequencedTaskRunner> embedder_task_runner) {
+  if (model_path.empty()) {
+    return std::unique_ptr<TextEmbedder, base::OnTaskRunnerDeleter>(
+        nullptr, base::OnTaskRunnerDeleter(embedder_task_runner));
+  }
+  auto embedder = std::unique_ptr<TextEmbedder, base::OnTaskRunnerDeleter>(
+      new TextEmbedder(model_path, embedder_task_runner),
+      base::OnTaskRunnerDeleter(embedder_task_runner));
+  return embedder;
+}
+
+TextEmbedder::TextEmbedder(
+    const base::FilePath& model_path,
+    scoped_refptr<base::SequencedTaskRunner> embedder_task_runner)
+    : model_path_(model_path),
+      owner_task_runner_(base::SequencedTaskRunner::GetCurrentDefault()),
+      embedder_task_runner_(embedder_task_runner) {
+  DCHECK(owner_task_runner_->RunsTasksInCurrentSequence());
+}
+
+TextEmbedder::~TextEmbedder() = default;
+
+bool TextEmbedder::IsInitialized() const {
+  return tflite_text_embedder_ != nullptr;
+}
+
+void TextEmbedder::Initialize(base::OnceCallback<void(bool)> callback) {
+  DCHECK(owner_task_runner_->RunsTasksInCurrentSequence());
+
+  embedder_task_runner_->PostTask(
```

Comment:
why? I'd say just use `base::SequenceBound<TextEmbedder>`. should be safer and simpler.

Reply:
We can't because `TextEmbedder` owns `tflite::task::text::TextEmbedder`.
If we force to delete `TextEmbedder` even on the correct sequence, it will hit its own code [failed std::vector assertion](https://source.chromium.org/chromium/chromium/src/+/main:third_party/tflite/src/tensorflow/lite/core/interpreter.h;l=787;drc=a56df8e74cfd73005acd6ffc41925d268152d3fb): `failed: front() called on an empty vector`.
But if we wait for it to finish on the right sequence, that task runner will block shutdown because it takes time to finish all of the `tflite::task::text::TextEmbedder::Embed` calls.
So we have to do [tflite::task::core::BaseTaskApi::Cancel](https://source.chromium.org/chromium/chromium/src/+/main:third_party/tflite_support/src/tensorflow_lite_support/cc/task/core/base_task_api.h;l=95;drc=82dff63dbf9db05e9274e11d9128af7b9f51ceaa;bpv=1;bpt=1) on a different thread to interrupt it as its comment indicates `// Usually called on a different thread than the one inference is running on.` The same technique is used by upstream `TFliteModelExecutor` to interrupt long running task when [timed out](https://source.chromium.org/chromium/chromium/src/+/main:components/optimization_guide/core/tflite_model_executor.h;l=550;drc=82dff63dbf9db05e9274e11d9128af7b9f51ceaa;bpv=1;bpt=1) to safely terminate the tflite model execution.

Reply:
I will try to implement it with what you suggested. If it turned out somehow not working, I will add locking mechanism for `tflite_text_embedder_`

Reply:
> Right now you're settings/getting it from different threads without any locking mechnism

I'm confused here, it's always on the same `embedder_task_runner` and callback is bound to the same thread who own `TextEmbedder`

Reply:
>  so tsan will be right to complain here

I built a tsan build on linux x64 and I didn't see any warnings/errors when running test scenario.

Reply:
> I think it's possible to use base::SequenceBound<std::unique_ptr<TextEmbedder>>. This way you can construct the instance manually, create a "CancelClosure" and store both text_embedder_ and text_embedder_cancel_closure_ in ConversationDriver.

Actually we won't be able to run `tflite_text_embedder_->Cancel()` in CancelClosure on different sequence with `base::SequenceBound<std::unique_ptr<TextEmbedder>>` which would always run on `embedder_task_runner` so I still don't think `base::SequenceBound` works in this case.

Reply:
I added lock with `GUARDED_BY` for `tflite_text_embedder_`  in 18fc1c70755447dc49d4e095743199c81739fcc3

Reply:
ok, right now it doesn't look right, because you're holding the lock while running the `Embed` call. This will block the call to `Cancel` until `Embed` finishes.

let's go without the lock for now (remove the last commit).

Maybe it's not that scary, i.e. reading a raw pointer should be atomic on modern architectures, so we should be fine.

I'm surprised that `tsan` didn't complain here though.

Reply:
Thanks, fixed that in https://github.com/brave/brave-core/pull/24953/commits/31d62adadc335b958f775cd765fbe0543f3cb405

Reply:
This class looks like it has some serious issues to me. It's getting called on more than one sequence which is fundamentally unsafe and makes it essentially impossible to safely destroy. It's using a lock in some places for `tflite_text_embedder_`, but not others. I think there are some major problems here and if you find yourself calling a single object on more than one thread you should definitely rethink your design.

Reply:
The only way to make calling a single class on more than one sequence safe is to use ThreadSafeRefCounted, otherwise the destruction is racy.

Reply:
I guess TextEmbedder can be safe because it's only called externally on a single thread by the owner and it gets destroyed on the embedder thread, but `tflite_text_embedder_` is currently not safe and it's still better to split this up so all the calls on the embedder sequence go to an object that lives exclusively on the embedder sequence so it can have proper guards and is easy to reason about.

Reply:
maybe `tflite_text_embedder_` is ok because `TextEmbedder` gets destroyed on the embedder thread so `tflite_text_embedder_` does as well, but it's all way more difficult to verify than it should be

Reply:
I will make changes to split them up.

Reply:
https://github.com/brave/brave-browser/issues/41645